core/arch/arm/mm/tee_pager.c

```diff
@@ -1280,30 +1288,49 @@ bool tee_pager_handle_fault(struct abort_info *ai)
 		 * address instead of the real virtual address.
 		 *
 		 * The i-cache can also be PIPT, but may be something else
-		 * to, to keep it simple we invalidate the entire i-cache.
-		 * As a future optimization we may invalidate only the
-		 * aliased area if it a PIPT cache else the entire cache.
+		 * too like VIPT. The current code requires the caches to
+		 * implement the IVIPT extension, that is:
+		 * "instruction cache maintenance is required only after
+		 * writing new data to a physical address that holds an
+		 * instruction."
+		 *
+		 * To portably invalidate the icache the page has to
+		 * be mapped at the final virtual address but not
+		 * executable.
 		 */
 		if (area->flags & (TEE_MATTR_PX | TEE_MATTR_UX)) {
+			uint32_t mask = TEE_MATTR_PX | TEE_MATTR_UX |
+					TEE_MATTR_PW | TEE_MATTR_UW;
+
+			/* Set a temporary read-only mapping */
+			area_set_entry(pmem->area, pmem->pgidx, pa,
+				       attr & ~mask);
+			tlbi_mva_allasid(page_va);
+
 			/*
 			 * Doing these operations to LoUIS (Level of
 			 * unification, Inner Shareable) would be enough
 			 */
-			cache_op_inner(DCACHE_AREA_CLEAN, pmem->va_alias,
-					SMALL_PAGE_SIZE);
-			cache_op_inner(ICACHE_INVALIDATE, NULL, 0);
+			cache_op_inner(DCACHE_AREA_CLEAN, (void *)page_va,
+				       SMALL_PAGE_SIZE);
+			cache_op_inner(ICACHE_AREA_INVALIDATE, (void *)page_va,
+				       SMALL_PAGE_SIZE);
```

Comment:
I wonder if the old page content was invalidated from icache in case it used to contain instruction ?